### ISSUE TITLE
Test crawler performance

### ISSUE THREAD

**Issue by SebastianZimmeck** [privacy-tech-lab/privacy-pioneer-web-crawler#9]
Before we start the crawl, we need to test the crawler's performance. So, we need to compare the manually observed groundtruth with the analysis results. We probably need a 100-site test set. 

- How do we select the test set sites given the different locations and states (issue #7) so that we have good test coverage?
- One issue is that different loads of a site may lead to different trackers etc. detected. So, we need to look for the groundtruth and analysis results at exactly the same site load. So, maybe, just load one site, get both groundtruth and analysis results and check?
- We need to document all of that

(@JoeChampeau and @jjeancharles feel free to participate here as well.)

**Comment by SebastianZimmeck** [comment]
Where are we with the testing protocol, @danielgoldelman?

**Comment by danielgoldelman** [comment]
Preliminary testing protocol

1. Run the crawl, collect the data. 2. Separate the pp data from the entries data.  For pp data: 1. Create spreadsheet for each root url 2. Log every piece of data into the spreadsheet with everything pp gives us, separated by pp data type   For all http request data: 1. Create spreadsheet for each root url 2. Do the most generic string matching with the values we are looking for. Note: we will have lists of keywords per vpn, we can get the ipinfo location while using the vpn by going to their site, and we can find monetization labels within the http requests. EX: if the zip code should be 10001, instead of a regex of \D10001\D, we look for just the string 10001. For every single key we could be looking for, we run it on the http requests gathered. Coallate these possible data stealing requests 3. Go through every http request and label, adding to the spreadsheet when necessary  Now to bring both together: 1. We have the two spreadsheet documents now. Time to classify 2. Potentially in a new spreadsheet, place all http requests that occur in both pp and all http requests first. Then all that only occurred in pp. then all that only occurred in the http requests.
3. Perform classification

**Comment by SebastianZimmeck** [comment]
@danielgoldelman, can you reformat and reorder [your comment](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-1886031606)?  The order is very hard to follow, there are multiple numbers 1 and 2 after each other, etc.

**Comment by danielgoldelman** [comment]
@SebastianZimmeck sorry, the original comment was written on GitHub mobile, so formatting was hard to check. Changes made above.

**Comment by SebastianZimmeck** [comment]
@danielgoldelman and @dadak-dom, please carefully read section 3.2.2 and 3.3 (in particular, 3.3.2) of [our paper](https://sebastianzimmeck.de/zimmeckEtAlPrivacyPioneer2024.pdf). We can re-use much of the approach there. I do not think that we need an annotation of the ground truth, but both of you should check the ground truth (for whatever our definition of ground truth is) and come to the same conclusion.

We have to create a testing protocol along the lines of the following:

1. Select the set of [analysis functionality](https://github.com/privacy-tech-lab/privacy-pioneer#7-privacy-practice-analysis) that we are testing and how
   - By default all analysis functionalities
   - But how are we going to test for keywords, for example? How for email addresses, how for phone numbers, ...? 

2. Pick a set of websites to test
   - How many? Probably, 100 to 200. We need some reasonable standard deviation. For example, it is meaningless to test a particular analysis functionality for just one site because a successful test would not allow us to extrapolate and claim that we are successful for, say, 1,000 sites in our crawl set with that functionality. So, we need, say, 10 sites successfully analyzed to make that claim. Can you solidify that? What is the statistical significance of 10 sites? @JoeChampeau can help with the statistics. We should have some statistical power along the lines of "with 95% confidence our analysis of latitude is within the bounds of a 10% error rate" (e.g., if we detect 1,000 sites having a latitude with 95% confidence the real result is between 900 and 1,100 sites).
   - Which sites to select? Again, the selected set should allow us to make the claim that if an analysis functionality works properly on the test set, it also works for the large set of sites that we crawl. So, we would need to pick a diverse set of sites covering [every analysis functionality](https://github.com/privacy-tech-lab/privacy-pioneer#7-privacy-practice-analysis) for each region that we cover. There should be no bias. For example, there will be no problems for monetization categories because they occur so frequently, but how do we ensure, e.g., that there is a meaningful number of sites that collect latitudes? Maybe, pick map sites from somewhere? How do we pick sites for keywords (assuming we are analyzing keywords)?
   - Are we using the same test set of sites for each country/state? Yes, no, is some overlap OK, is it harmful, is it good, ...?
   - How are we selecting sites randomly? Use random.org.
   - We can't select any sites that we used for preliminary testing, i.e., validation. So, which are the sites, if any, that need to be excluded? If we randomly select an excluded site, how do we pick a new one? Maybe, just the next one on a given list.

3. Running the test
    - Are we testing one site a time or run the complete test set? If we do the former we need to record all site data (and be absolutely sure that there are no errors and nothing omitted in the recording). We need to get both the analysis results _*and*_ the ground truth data at the same time. The reason is that when we load a site multiple times, there is a good chance that not all trackers and other network connections are identical for both loads. So, the analysis results could diverge from the ground truth if the latter is based on a different load. We need to check the ground truth for the exact site load from which we got the analysis results. The alternative to a complete test set crawl is to do the analysis for one site at a time, i.e., visit a site, record the PP analysis results, use browser developer tools (and other tools, as necessary) to check the ground truth, record the evidence, record the ground truth evidence and result, then analyze the next site and so on. So, we would be doing multiple crawls of one site.
    - We will also need to change the VPN for every different location.
    - Who is going to run the test? @JoeChampeau has the computer. Is it you, @dadak-dom or @danielgoldelman? Both the PP analysis results and the ground truth should be checked by two people independently. This seems easier if only one test set crawl is done as opposed to the site-by-site approach.

4. Ground truth analysis
   - How do we analyze the ground truth? Per [your comment above, @danielgoldelman](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-1886031606), I'll take that we do string matching in HTTP messages. Is that a reliable indicator? Maybe, we would also need to look at, say, browser API usage for latitude, i.e., the browser prompting the user to allow location access. What are the criteria to reliably analyze the ground truth? This can be different for our different functionalities.

These question cannot be answered in the abstract. @danielgoldelman and @dadak-dom, please play around with some sites for each analysis functionality and come up with a protocol to analyze it. For each functionality you need to be convinced that you can reliably identify true positives (and exclude false positives and false negatives). In other words, please do some validation tests.

**Comment by dadak-dom** [comment]
> Who is going to run the test?

Would it make sense if @JoeChampeau runs the test, and then hands the data over to Daniel and me? I thought it would make sense since that's the computer that we will use to run the actual crawl. That way, we could avoid any potential issues arising when switching between windows and mac. Just a thought.

@SebastianZimmeck , the way I understand it, we will end up with three different site lists for each country (please correct me if I'm wrong)

1. Validation (what Daniel and I are doing now)
2. Test set (what we're preparing for and will soon be running)
3. The actual crawl list.
We cannot have any overlap between the validation and the test set, but can the test set (and/or the validation) be derived from the actual crawl list? I would need to know this before I start making any lists for the test set.

**Comment by SebastianZimmeck** [comment]
> Would it make sense if @JoeChampeau runs the test, and then hands the data over to Daniel and me?

It certainly makes sense, but that would depend on if @JoeChampeau has time as the task was originally @danielgoldelman's. (Given our slow speed, the point may more or less resolve itself since we will be all back on campus soon anyways.)

> (please correct me if I'm wrong)

All correct.

> but can the test set (and/or the validation) be derived from the actual crawl list?

Yes, the validation and test set can be derived from the crawl list.

**Comment by dadak-dom** [comment]
I have added my proposed crawl testing lists to the branch connected with this issue (issue-9). Here was my procedure:

1. For each country that we will crawl, create a new .csv file.
2. Go to random.org and have it generate a list of random integers from 1-525
3. Take the first six integers and find the matching URL from the general list
4. Regenerate the random integers and find the six matching URLs from the country specific list
5. If there seems to be a bias for one functionality, throw the list out and try again. (Or, if there is any overlap with sites that were used for validation. Luckily, this was never the case for me)
6. Repeat the process for each location we will crawl, so ten times total.

With point 5 I tried my best to include a fair share of sites that take locations, as monetization was easy to come by. @SebastianZimmeck let me know if any changes need to be made.

**Comment by SebastianZimmeck** [comment]
OK, sounds good!

So, our test set has a total of 120 sites? For each of the 10 countries/states 6 sites from the general list and 6 from the country-specific list.

> With point 5 I tried my best to include a fair share of sites that take locations

How did you make the guess that a site takes locations?

**Comment by dadak-dom** [comment]
> So, our test set has a total of 120 sites? For each of the 10 countries/states 6 sites from the general list and 6 from the country-specific list.

Yes, 120 sites total. 

> How did you make the guess that a site takes locations?

A couple of ways, e.g. visiting the site and seeing if it requests the location from the browser, or if PP detects a location, or if I know from my own browsing that the site would take locations.

**Comment by SebastianZimmeck** [comment]
OK, sounds good!

Feel free to go ahead with that test set then. As we discussed yesterday, maybe the performance is good. Otherwise we call that set a validation set, pick a new test set, and repeat the test (after fixing any shortcomings with the crawler and/or extension).

One important point, the PP analysis needs to be set up exactly as it would be in the real crawl, i.e., with VPN, a crawl not just the extension. Though, it does not need to be on the crawl computer.

**Comment by dadak-dom** [comment]
One more thing: I noticed this morning that there are a lot of sites in the general list that redirect to sites that are already on the list. Can't believe I didn't catch that sooner, so I'll fix that ASAP. Just to be safe, I'll also redo the general list part of the test set.

**Comment by SebastianZimmeck** [comment]
Great!

**Comment by dadak-dom** [comment]
@SebastianZimmeck I'm compiling the first round of test data, but so far I'm not getting as many location requests found as I'd like. You mention in one of the comments above that it might be worthwhile to make a list of, say, map sites. If I were to make a test list of sites with the clear intention of finding location requests, how can I make it random? Would it be valid to find, for example, a list of 200 map sites (not necessarily from the lists that we have), and pick randomly from that? If not, what are some valid strategies?

**Comment by SebastianZimmeck** [comment]
> what are some valid strategies?

Just map sites would probably be too narrow of a category. There may be techniques that are map site-specific. In that case our test set would only claim that we are good at identifying locations on map sites. So, we need more categories of sites, ideally, all categories of sites that typically get people's location. 

Here is a starting point: [Can you give some examples of websites that use geolocation to target local customers?](https://www.quora.com/Can-you-give-some-examples-of-websites-that-use-geolocation-to-target-local-customers) So, the categories mentioned there, plus map sites, plus any other category of site that you found in your tests that collect location data. Maybe, there are generic lists (Tranco, BuiltWith, ...) that have categories of sites. Compile a list out of those and then randomly pick from them. That may be an option, but maybe you have a better idea.

So, maybe our test set is comprised of two parts:

1. Location test set
2. Monetization and Tracking test set

Maybe, it even has three parts if tracking pixel, browser fingerprinting, and/or IP address collection (the Tracking categories) are also rare. Then, we would also need to do a more intricate test set construction for the Tracking categories as well. I would expect no shortage of sites with Monetization.

There are no hard rules for testing. The overall question is: 

**What test would convince you that the crawl results are correct? (as to lat/lon, IP address, ... )** 

What arguments could someone make if they wanted to punch a hole in our claim that the analysis results demonstrate our crawl results are correct? Some I can think of: too small test set, not enough breadth in the test set, i.e., not covering all the techniques that we use or types of sites, sites not randomly selected, i.e., biased towards sites we know work ... (maybe there are more).

I would think we need at least 100 sites in the test set overall and generally not less than 10 sites for each practice we detect (lat/lon, tracking pixel, ...). Anything less has likely not enough statistical power and would not convince me.

**Comment by dadak-dom** [comment]
I've just added the lists and data that we can use for the first go at testing. A couple things to note:

1. I managed to get a set where PP detected at least 10 of nearly every analysis functionality we were looking for, except for Zip Code and Lat/Long. My theory is that using the VPN makes it harder for sites to take this information, and so there's no requests with this information for PP to find. Of course, this will only be verified after testing fully, but I wanted to raise the possibility that these two analysis functions may not be possible with the setup we are going with. Just from the amount of sites, it's strange that none of them took lat/long, and yet many took region and city. I also did a quick test where I found a site I knew would take lat/long or zip code, and visited it without a vpn to make sure PP found those things. I then connected to the same site with a VPN, and PP wouldn't find lat/long or zip, but it still found Region and City. The good news is that Region and City seem to pop up quite a bit, so I believe we should have no problem testing for them.
2. For documentation, here was my procedure for generating the lists:

- I had two pools of sites: one was a mixture of the top sites of different categories, as well as sites that I had encountered in my personal browsing. The other set was the top 100 sites from the builtwith list.
- For each list, generate a set of random integers. Use the corresponding row number for the site that you'll select
- Take six sites from the mixture, and six sites from the pre-compiled list, for each country that we crawl.
- In theory, you should have 12 sites. However, a few of them were bound to crash, and so as long as fewer than two crashed for a given crawl, I thought it was ok, since we'll still have over 100 sites. So you should have 10-12 sites per country/state.
- If you generate a random integer that corresponds to a URL that was already taken, use the next available URL.

When crawling, I made sure that I was connected to the corresponding VPN for each list, i.e. when crawling using the South Africa list, I was connected to South Africa.

**Comment by SebastianZimmeck** [comment]
Good progress, @dadak-dom!

> I then connected to the same site with a VPN, and PP wouldn't find lat/long or zip, but it still found Region and City. 

Not having lat/long would be substantial. Can you try playing around with the Mullvad VPN settings?

<img width="320" alt="Screenshot 2024-01-17 at 10 06 50 PM" src="https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/assets/11034003/ff11afac-deff-485e-af6b-f59debf7efda">

Can you try allowing as much as possible? Our goal is to have the sites trigger as most as possible of their tracking functionality.

Also, while I assume that the issue is not related to Firefox settings since you get lat/long with presumably the same settings in Firefox with VPN and Firefox without VPN, we should also have the Firefox settings as allowing as much as possible.

Maybe, also try a different VPN. What happens with the Wesleyan VPN, for example?

The bottom line: Try to think of ways to get the lat/long to show up.

**Comment by dadak-dom** [comment]
I messed around with the settings for both Firefox Nightly and Mullvad, no luck there.

I've tried crawling and regularly browsing with both Mullvad and the Wesleyan VPN. I was able to get Wesleyan VPN to show coarse location when browsing, but not when crawling. Under Mullvad, coarse/fine location never shows up. 

However, when trying to figure this out, I noticed something that may be of interest. Per the Privacy Pioneer readme, the location value that PP uses to look for lat/long in HTTP requests is taken from the Geolocation API. Using the developer console, I noticed that this value doesn't change, regardless of the location that you are using for a VPN.
Maybe something strange is going on with my machine, so to check what I did, I encourage anyone to try the following:

1. Without a VPN connection, visit any website.
2. Paste in the following code into your developer console: 

const options = {
  enableHighAccuracy: true,
  timeout: 5000,
  maximumAge: 0,
};

function success(pos) {
  const crd = pos.coords;

  console.log("Your current position is:");
  console.log(`Latitude : ${crd.latitude}`);
  console.log(`Longitude: ${crd.longitude}`);
  console.log(`More or less ${crd.accuracy} meters.`);
}

function error(err) {
  console.warn(`ERROR(${err.code}): ${err.message}`);
}

navigator.geolocation.getCurrentPosition(success, error, options);

3. Compare this value to what ipinfo.io gives you by visiting [ipinfo.io](https://ipinfo.io/) (without a VPN, they should be roughly the same)
4. Now do steps 2 and 3 while connected to a VPN in a different country

When I do these steps, I end up with a different value for ipinfo, but the value from geolocation API stays the same (the above code should be set to not use a cached position, i.e. maximiumAge: 0)
I then looked at location evidence that PP collected for crawls I did when connected to other countries. Sure enough, PP would find the region and city, because that info is provided by ipinfo. However, PP would miss the lat/long that was in the same request, most likely because the geolocation API is feeding it a different value, and so PP is looking for something else.

However, this doesn't explain why PP doesn't generate entries for coarse and fine location when crawling without a VPN. From looking at the ground truth of some small test crawls, there clearly are latitudes and longitudes of the user being sent, but for some reason PP doesn't flag them. @danielgoldelman , maybe you have some idea as to what is going on? This doesn't seem to be a VPN issue as I initially thought.

**Comment by danielgoldelman** [comment]
Interesting. I was having different experiences, @dadak-dom ... lat and lng seemed to be accurately obtained when performing the crawls before. Have you modified the .ext file?

**Comment by dadak-dom** [comment]
No, I didn't make any changes to the .ext file, @danielgoldelman . Was I supposed to?

**Comment by SebastianZimmeck** [comment]
Good progress, @dadak-dom!

> Using the developer console, I noticed that this value doesn't change, regardless of the location that you are using for a VPN.

> When I do these steps, I end up with a different value for ipinfo, but the value from geolocation API stays the same 

Hm, is this even a larger issue not related to the VPN? In other words, even in the non-VPN scenario do we have a bug that the location is not properly updated? This is the first point we should check. (Maybe, going to a cafe or other place with WiFi can be used to get a second location to test.)

What is not clear to me is that when we crawled with different VPN locations for constructing our training/validation/test set, we got instances of all location types. So, I am not sure what has changed since then.

@danielgoldelman, can you look into that?

**Comment by dadak-dom** [comment]
I forgot to use the hashtag in my most recent commit, but @danielgoldelman and I seem to have solved the lat/long issue. Apparently the browser that selenium created did not have the `geo.provider.network.url` preference set, and so the extension wasn't able to evaluate a lat or long when crawling. My most recent commit to issue-9 should fix this, but this should be applied to the main crawler as well. Hopefully, this means that we can get started with gathering test data and testing.

**Comment by danielgoldelman** [comment]
Additionally, we have run the extension as if we were the computer, and compared our results for lat/lng with what we would expect the crawl to reasonably find. This approach worked! We used the preliminary validation set we designated earlier on, so this claim should be supported via further testing when we perform the performance metric crawl, but on first approach the crawl is working as intended for lat/lng.

**Comment by SebastianZimmeck** [comment]
Great! Once you think the crawler and analysis works as expected, feel free to move to the test set.

**Comment by danielgoldelman** [comment]
Rebased the main branch of PP into the crawler branch of PP to ensure that we are working with the correct extension when we perform the crawls

**Comment by SebastianZimmeck** [comment]
@danielgoldelman will come up with the testing protocol and together with @dadak-dom and @natelevinson10 (and possibly @JoeChampeau) perform the test.

**Comment by JoeChampeau** [comment]
@danielgoldelman and I are resolving the final issues with the crawler right now. Once we're done, @danielgoldelman will run the test crawl using the test lists and then we'll have those results.

**Comment by dadak-dom** [comment]
Finally, some good news regarding testing. As I brought up in last week's discussion, there was an issue where zip codes were not being identified properly when crawling. I've done some work on that front, and that's been fixed. It looks like Privacy Pioneer is finally working on the crawler (🥳). 
On another note, I re-worked the test list, as well as the methodology to create it, so I'll document that process here. 

To create the 100 site test list, I first gathered a list of 10 different APIs that seem to take location data, whether via my own browsing or through their descriptions on BuiltWith. Then, per Kate's suggestion, I used the lists of live sites that BuiltWith claims use said APIs to construct the test list. I did this by randomly selecting 10 sites from each list, using random.org. If a randomly selected site was a redirect / faulty in some way, I used the next randomly generated number. I chose 10 different APIs so that we weren't biasing too hard towards any one format of HTTP request. Additionally, I wouldn't be too concerned over bias because (a) not every site chosen from the lists uses the API in the same way, and (b) some of the sites chosen use more than one API for geolocation. Ultimately, in order to get enough examples of location data being taken, some bias needed to occur (past experience tells us that simply selecting random sites tends to not yield that much location evidence), though I believe that the steps I've taken have found a good compromise. 

Here are the lists in question, each one having been accessed on April 24, 2024:
https://trends.builtwith.com/websitelist/AB-Tasty
https://trends.builtwith.com/websitelist/Intellimize
https://trends.builtwith.com/websitelist/Dynamic-Yield
https://trends.builtwith.com/websitelist/Permutive
https://trends.builtwith.com/websitelist/securiti
https://trends.builtwith.com/websitelist/IPinfo
https://trends.builtwith.com/websitelist/ipdata
https://trends.builtwith.com/websitelist/IP-API
https://trends.builtwith.com/websitelist/Ipregistry
https://trends.builtwith.com/websitelist/Rebuy
(I will also put up the downloads of the list onto the Google Drive)

**Comment by SebastianZimmeck** [comment]
Excellent, @dadak-dom! Let's discuss the details tomorrow.

**Comment by SebastianZimmeck** [comment]
A few thoughts on how we test the performance of Privacy Pioneer's analysis when used in a web crawl and with a VPN.

The starting point is that we know Privacy Pioneer's analysis performance when used without a crawler and without a VPN. We measured that in our paper. Now, the two elements --- crawler and VPN --- can introduce errors. How can we measure that? To which extent should we measure that?

As I see it at the moment, there are two points:

1. We can analyze Privacy Pioneer's performance with crawler and VPN just in the normal way. In other words, we can connect to a VPN, run the crawl, capture the analysis results, capture all data that Privacy Pioneer had available, and compare the analysis results against the results of just Privacy Pioneer running on the data. There needs to be some manual checking, but that should be ultimately easy to do. This is what @danielgoldelman and @dadak-dom were describing.
2. The harder question, about which I am not sure, is to which extent we should also measure Privacy Pioneer with crawler and VPN against Privacy Pioneer without those, i.e., in both instances Privacy Pioneer will necessarily need to run on different datasets necessarily causing different results. At the moment, I am leaning towards doing such analysis because we know from @danielgoldelman's findings that there are differences in the data when using a crawler and VPN. So, if we claim that we capture a realistic situation a user would encounter in a particular country, we would need to say that with VPN and crawl is more or less equal to without VPN and crawl.

On the second point, I also find it plausible to say that with VPN and crawl the situation is a lower bound. That is at least what people are exposed to in real life. But this assumes that crawl and VPN leads to just _less_ data as opposed to _incorrect_ data. So, that speaks for doing the second analysis to identify incorrect results.

Given that, the issue for the second analysis becomes that every run of the test set is different. So, a site may load different ad networks or load them at different times, for example. Thus, comparing two different runs will often result in two different result sets even when no error occurs. I am not sure to which extent we can distinguish incorrect from just fewer results (e.g., VPN introduced wrong location is probably more incorrect than just omitted).

One suggestion could be:
- Run Privacy Pioneer on the test set without VPN and crawler and capture the results
- Run Privacy Pioneer on the test set without VPN and crawler and capture the results a second time. This second run will give a sense of the delta between two different runs (there could be even a third, fourth, etc. run; the more runs, the clearer we understand the variance between different runs)
- Run Privacy Pioneer on the test set with VPN and crawler _with VPN set to our real location_  and capture the results and compare those against the previous runs without VPN and crawler (possibly, also do this run multiple times). In essence, does this run look similar in terms of variance of error rates compared to the previous runs without crawl and VPN? If so, we are good.

Having fewer analysis results as a matter of having less data to analyze is not an issue if we say we are aiming for a lower bound. However, having incorrect analysis results would be problematic.

Also, I think we would not need to test (extensively) for different VPNs (i.e., run Privacy Pioneer on the test set with VPN does not necessarily mean run on every country/state VPN). If we have reason to believe that it works for one VPN location, from a systematic viewpoint, there is no reason to mistrust the other VPN locations as they follow the same systematic. This would only leave open the possibility that there is a location-specific problem with one particular VPN. I think that would be proper to state as an assumption. We could also try running on different VPN locations to see how it goes, though.

Any thoughts? Is this thinking correct? Other ideas?

**Comment by dadak-dom** [comment]
You raise some good points, @SebastianZimmeck . Nate, Daniel, and I discussed this a little, and we had some thoughts that we'd like your opinion on. From your comment, it seems like there's two approaches that make the most sense to me.

The first option is essentially what you were describing: We take the test set that I constructed and perform a crawl both on the Cloud VM  as well as locally. We then shut down the VM, turn it on again, and run the test list again, performing the crawl 3 times for each machine (local and virtual). By repeating the crawl, we mitigate the chances of anomalies in our test data that occurred (site temporarily down, etc.). Performing these test crawls would give us insight into how the VM infrastructure affects the evidence that Privacy Pioneer finds.

The second option would also entail 3 test crawls, but instead of crawling locally, someone manually clicks through the test list (preferably a subset, since this would take much longer than option one). Both the actual crawl and the manual "crawl" would be performed on the VM. This would give us insight into the effect that the crawler infrastructure has on the evidence that Privacy Pioneer finds. Alternatively, we could perform the manual crawl locally and verify both points (effect of VM, effect of crawl) at once. Once we gather this data, we should have a better idea of whether or not crawling is a good idea. I could take the lead on the manual crawl if that's the route that we go with, asking for help as needed. 

What are your thoughts on this, @SebastianZimmeck ? (or anyone?)

**Comment by SebastianZimmeck** [comment]
Essentially we need to compare:

`Privacy Pioneer` vs `Privacy Pioneer + VM + Crawl infrastructure`

So, there are two error sources:
1. VM
2. Crawl infrastructure

It certainly would be more systematic to evaluate each source on its own (e.g., to pinpoint error rates to the sources).

Your second option, @dadak-dom, is going in the systematic direction. Here is a variation:
- `Manual crawl without VM` vs `manual crawl on VM` -> evaluates impact of crawl component (i.e., manual crawl == normal Privacy Pioneer usage)
- `Real crawl locally` vs `real crawl on VM` -> evaluates impact of VM

Would that make sense? Maybe, do each three times to account for normal fluctuation of loaded site elements. 

Generally, what you are saying makes a lot of sense to me, @dadak-dom.

**Comment by dadak-dom** [comment]
> Here is a variation:

Your suggestion makes sense to me, @SebastianZimmeck . If you think it's alright, I'll start collecting the testing data as soon as possible 👍

**Comment by SebastianZimmeck** [comment]
Sounds good! Feel free to go ahead, @dadak-dom! And we can also discuss further in our meeting this week.

**Comment by SebastianZimmeck** [comment]
As discussed, @dadak-dom will begin to evaluate Privacy Pioneer vs Privacy Pioneer VM.

**Comment by dadak-dom** [comment]
I have essentially finished the comparisons between Privacy Pioneer and Privacy Pioneer with a VM. Along with my results, I'm going to detail how I went about this analysis, and my reasoning for setting it up in this way. I think this will help get @atlasharry up to speed, and also hopefully will clear up any confusion that anyone may have.

### Goals
With this analysis, I set out to uncover whether or not introducing a VM into our web crawl has a significant impact on any data that we would collect, as brought up [here](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-2081560814). Ideally, we would want to see that, for every category, the number of requests flagged by Privacy Pioneer on a VM is either the same or lower than what would be discovered via the control group (no VM and no crawler). 

### Setup / Procedure
Based on the goals outlined above, I created a relatively simple, although time-consuming, procedure. From a bird's eye view, I essentially posed as a regular user, manually clicking through the [100-site test list](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/blob/main/selenium-crawler/test-list3.csv) I had created earlier. Here's the process in a little more detail:
(Preliminary steps: Make sure that you have cloned into both the crawler _and_ the extension repo. Also make sure that you set the [boolean flags](https://github.com/privacy-tech-lab/privacy-pioneer/blob/main/src/background/background.js) that put Privacy Pioneer into crawl mode to true. This way, we can overwrite both the incorrect data that IPinfo sends, as well as the lat/long coordinates on the cloud)
1. Identify a local machine to use (i.e., my laptop). 
2. Boot up the VM from the location that we are testing against.
3. Both locally and on the cloud, launch _two_ instances of the terminal
4. In the first instance of the terminal, launch Privacy Pioneer in development mode, as described [here](https://github.com/privacy-tech-lab/privacy-pioneer?tab=readme-ov-file#3-development.)
5. In the second instance of the terminal, launch the [rest-api](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/tree/main/rest-api).
6. Now that Privacy Pioneer has been opened with a new instance of Firefox, we can input the location data that we need to overwrite. For the local machine, just use your real location. For the VM, look up your coordinates and zip code based on the city that the VM is based in. 
7. Now, we can start "crawling". Copy-paste the first URL from the test list (both locally and virtually).
8. Wait until the URL is printed by the terminal running the rest-api, which should take around a minute. This indicates that data has been recorded, and we're ready to move on to the next site.
9. Repeat steps 7-9 until the entire list has been "crawled".
10. In case of any crashes, delete any data gathered from the last site, restart Privacy Pioneer and the rest-api, and start crawling again from where you left off.

After running this process three times, we should have plenty of data for each of the 100 sites. However, since Privacy Pioneer records data in JSON format, I needed a way to meaningfully import the data I collected into a spreadsheet program. Thus, I created a short Python script that would take in all of the individual JSON entries and create a CSV file detailing how many pieces of evidence were acquired for each category for every site. [Here's an example](https://docs.google.com/spreadsheets/d/1fHwTsZ_9Vrg4OHT4C9ONn4aR_SgS6u77sESx52uk31o/edit#gid=1885569146).

### Results
Since this is essentially an experiment involving matched pairs, I thought it would make sense to analyze my findings using a paired t-test. I would first find the average number of requests per category per website. In other words, how many advertising/ipAddress/whatever requests would I expect from example.com when "crawling" locally? How about on a VM? Then, I would run a paired t-test, treating the local data as a control group and the VM data as the treatment group to find a p-value for every category.
After our last discussion, I also decided to run the experiment in two additional locations (Columbus and Iowa). Here are the overall results, with a [link](https://docs.google.com/spreadsheets/d/1fHwTsZ_9Vrg4OHT4C9ONn4aR_SgS6u77sESx52uk31o/edit#gid=701931879) to the entire Google Sheet if anyone is interested.
![image](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/assets/144378508/8ad4e863-6356-4d49-a7f2-59cfa163838d)

I have marked in green and red any stastically significant results. Based on this data, it seems like the only categories with any major differences to speak of are coarseLocation, fineLocation, region, and zipCode. For the first three elements, I think that these results are acceptable, since they indicate that there is statistical evidence for less data being generated on the cloud. It's not ideal, but we can work with it. The more interesting point, though, is zipCode, which seems to come up _more_ in the cloud for 2/3 locations. This may seem strange, but I have some possible explanations for it.

Since nearly all instances of location gathering are done through the IP, it's not unlikely that this is a shortcoming of where I'm crawling from, rather than Privacy Pioneer underperforming. I've noticed for a while now that, when I'm connected to the campus wifi, the zip code can switch between two different values. I've looked at the data that passed through Privacy Pioneer, and it seems like there's no real difference in the amount of zip codes being taken; what does change, however, is the zip code that websites are taking. Also, notice the averages in the categories for coarseLocation and zipCode. Virtually, the average number of requests per site for these two categories doesn't seem to differ by more than +/- 0.05. Locally, however, they differ by 0.2. This (and also my own experience of inspecting these requests) indicates that requests containing coordinate data oftentimes also include zip codes. These results, however, support my hypothesis of zip code fluctuation, since this would mean that certain instances of zip codes would go unnoticed, while coordinate data would still get uncovered, leading to the current results.

To try and remedy this, I could perform the local crawl again, this time from a different location with a static IP. I could try my local library, but I guess that depends on what you think I should do next, @SebastianZimmeck.

Overall, does this make sense? Am I going in the right direction? Is this like what you had in mind, @SebastianZimmeck? Would this convince you (or anyone else) that collecting this data on a VM is valid?

**Comment by SebastianZimmeck** [comment]
Nice work, @dadak-dom! 

As discussed, if you go a bit deeper, we will learn more about any inconsistent results.

**Comment by dadak-dom** [comment]
I have been looking into what exactly causes the inconsistencies between runs on a VM (without a crawler), and here's what I've found. Generally, there seem to be two main causes for why a certain run would have a different number of requests. The simplest explanation, which I've confirmed to be the case for at least a portion of the inconsistencies, is that different site loads result in different requests being sent. This is especially common with requests in the monetization category. Sometimes, I would be able to see that the request was supposed to come in, but something on the website's end would go wrong, resulting in the request erroring out. Additionally, certain requests would be counted in more than one category, so one request not showing up could potentially affect other categories as well. I think it's safe to say that for these types of inconsistencies, there's not much that we can do.

However, the location category seemed to have an additional inconsistency. I traced Privacy Pioneer's output in the developer console and I found that, for certain geolocation APIs, the ML model would return inconsistent results, e.g., on one site load a request wouldn't get flagged, while a second load would result in region and zip code getting flagged. This would happen even though the request snippets appeared identical. I'm not very familiar with the setup of the ML, so I'm not sure whether this behavior is expected or not. Could it be due to [this known issue?](https://github.com/privacy-tech-lab/privacy-pioneer-machine-learning#5-known-issues) I'd imagine it doesn't help. 

So, I've identified two sources of inconsistency when comparing one VM site visit to another. However, I'm a little confused about where I'm supposed to go from here. I'm not sure if either of these issues can really be "fixed" in a realistic time frame. Should I try to calculate the precision and recall for VM vs. Non VM to gauge whether or not there's a performance difference? Maybe just for the VM? Maybe there's a different direction I should take? Any input would be appreciated.

**Comment by dadak-dom** [comment]
Update: It looks like my initial thoughts regarding monetization / tracking were incorrect... I've been able to find instances where Privacy Pioneer failed to flag a request. I'm currently looking into diagnosing the issue with the extension. My first thoughts are that this is related to how Privacy Pioneer takes in the requests to begin with, but I'll update more as I get more answers.

**Comment by SebastianZimmeck** [comment]
OK, good that you confirmed.

**Comment by SebastianZimmeck** [comment]
@dadak-dom found that the likely reason for the discrepancy in location and other analysis results is the exclusion of a resource type in the Privacy Pioneer analysis. Per [@dadak-dom's PR](https://github.com/privacy-tech-lab/privacy-pioneer/pull/580):

> Turns out that our [HTTP request listener](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onBeforeRequest) was filtering out requests that were initiated using the [Beacon API](https://developer.mozilla.org/en-US/docs/Web/API/Beacon_API) (as opposed to XHR or Fetch). Certain requests would occasionally use Fetch on one run, but Beacon on the other, and so the extension would miss the latter requests completely.

@dadak-dom will do remaining testing to be sure and then add a small explanation in the [Known Issues section of the Privacy Pioneer readme](https://github.com/privacy-tech-lab/privacy-pioneer?tab=readme-ov-file#11-known-issues) that can link to a more detailed explanation here.

**Comment by SebastianZimmeck** [comment]
As discussed, @atlasharry will work on the creation of a 100-site test set list and @dadak-dom will investigate some final crawler inconsistencies.

**Comment by atlasharry** [comment]
I have created a new 100-site crawler test list and put it in our [Google Drive](https://drive.google.com/file/d/1JOVnpmUF9tLwil77ido8vgcgVg2jcTDQ/view?usp=drive_link). 
The list was created followed by the below procedures:
- First, I browsed through the [BuiltWith](https://trends.builtwith.com/analytics) and followed the trends tools of "Web Analytics Usage Distribution in the Top 1 Million Sites" to locate the possible websites that may use those tools to track the: "IP Address", "Tracking Pixels" "Corase Location“, ”Fine Location", "Fingerprint", "Region" and "City". I randomly gathered 50 websites from there.
-  Second, I followed [Tranco List](https://tranco-list.eu/) to randomly selected another 50 websites. 
- Third, I compiled the list and wrote a mini script to compare it with the original test lists to duplicate and make sure there is no same websites in both lists. 
- After the deduplication, I used a mini script to shuffle the site list randomly and uploaded it to the Drive.

**Comment by dadak-dom** [comment]
As I've been validating, I've also been working on refining a test procedure, which I'll link [here](https://docs.google.com/document/d/1OBT3AlhU-tCBpd1Jzh1TXoIX-cdlyk_i1K0AWWdYMew/edit). Without going into too much detail, the testing procedure is essentially me looking at the ground truth, identifying what values should have been positive results, and then comparing that against what Privacy Pioneer returned. I've done a more thorough write-up of what to look out for in the document linked above.

Additionally, I think I'm confident enough in the location gathering that I could get started on the test set. Before I get started, I have two main questions:
1. Is it necessary to label every single request that I'm analyzing? Based on the [GPC crawler testing](https://github.com/privacy-tech-lab/gpc-web-crawler/issues/54), it looks like they were mainly concerned with the counts of true positive, false positive, etc. An [older comment](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-1886031606) by Daniel mentions labeling requests individually, but I think this would add significant time to a task that is already going to take up a lot of it. If it's absolutely necessary, I can incorporate it.
2. What do we do with requests that are clearly instances of location gathering (i.e., I look at the request and it has taken my location data), but the model returns otherwise? Since we're only evaluating the extension + crawler + VM and not the model, my understanding is that, as long as Privacy Pioneer correctly takes in the request, sends it to the model, and receives a result, that that's the extension functioning as intended.

**Comment by SebastianZimmeck** [comment]
We move forward per our discussion.

One important point to remember is that @atlasharry will do some groundtruth analysis while in Seoul, which we can then compare to the Seoul VM.

**Comment by SebastianZimmeck** [comment]
Here is the [overall plan](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/49#issuecomment-2217874550). This issue is an omnibus testing issue. @dadak-dom has opened a separate issue for comparing the Crawler + VM results vs the local results (#49).

**Comment by dadak-dom** [comment]
Here are the results from my testing. As discussed, @atlasharry and I will look into possible explanations for lat/long and zipCode behaviors.
![image](https://github.com/user-attachments/assets/9987c5d9-3d0b-42af-936e-60df63b64178)

**Comment by dadak-dom** [comment]
Quick update from my end:
Initially, I was going to feed the false negative results from my test run on the VM into both Daniel's final model, and the model that I was tinkering with, with the expectation that the original Pytorch model would incorrectly classify the snippets, and maybe we'd see some improvements in the tuned model. Interestingly, feeding the Pytorch model with the mislabeled snippets actually resulted in correct results for nearly all the snippets I tried. In other words, there doesn't seem to be a problem with the original Pytorch model.
However, manually feeding the snippets into the TensorFlowJS model produced the same results as the test (false negative). I've confirmed with Daniel that there was no performance drop in converting from Pytorch to TensorFlow, so the only other thing I could think of is the [known issue](https://github.com/tensorflow/tfjs/issues/8025) regarding the conversion from TF to TFJS. I think my next step will be to try replicating Daniel's conversion and seeing if I can get anywhere with that.

**Comment by SebastianZimmeck** [comment]
> Interestingly, feeding the Pytorch model with the mislabeled snippets actually resulted in correct results for nearly all the snippets I tried. ... I've confirmed with Daniel that there was no performance drop in converting from Pytorch to TensorFlow, so the only other thing I could think of is the https://github.com/tensorflow/tfjs/issues/8025 regarding the conversion from TF to TFJS.

That is a very good point!

It is possible that the performance drop, at least some of it, has to do with the conversion from PyTorch/TensorFlow (in Python) to TensorFlow (in JavaScript). However, why are we seeing a bigger drop than we saw earlier?

From [our paper](https://sebastianzimmeck.de/zimmeckEtAlPrivacyPioneer2024.pdf):

PyTorch/TensorFlow (in Python) results:

<img width="432" alt="Screenshot 2024-08-03 at 1 13 53 PM" src="https://github.com/user-attachments/assets/0080d591-7a77-4ba3-bc0f-b14189001a40">

TensorFlow (in JavaScript) results:

<img width="451" alt="Screenshot 2024-08-03 at 1 14 28 PM" src="https://github.com/user-attachments/assets/a2b395c3-fe2f-4f5b-b9b3-3014493040d4">

For example, latitude dropped from a recall of 0.97 to 0.82. However, the current results [above](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-2231406248) show a drop to a recall of 0.5 (I am assuming fineLocation is latitude or longitude. As a side note, can we also test for latitude and longitude and not fineLocations as results for latitude and longitude can differ?)

The only reason I can think of why that is the case is that the current test set, more specifically, the instances of the current test set for which we get the false negative location results, are different compared to the original test set and/or we have more of those instances now. 

**So, can we run the current test set on the [version of the code at the time](https://github.com/privacy-tech-lab/privacy-pioneer/releases/tag/v1.2.1)?** 

If we are able to replicate @danielgoldelman's testing procedure with the code at that time but with our current test set, we should be getting results identical or at least close to what we are getting now with the new code and the current test set (if the conversion is the issue). 

I am making this suggestion as we discussed that it is difficult to replicate running the test set at the time on our new code. That would also be an option if possible. In that case the old code and new code should return identical or at least very similar results.

> I think my next step will be to try replicating Daniel's conversion and seeing if I can get anywhere with that.

That is also worthwhile to get a better understanding in general. Maybe, there is something we are currently not thinking of.

**Comment by SebastianZimmeck** [comment]
## 1. Current Status of the Model

It works! @atlasharry and @PattonYin made some great progress! As it turns out, there is nothing wrong with the model. Both the (1) PyTorch/TensorFlow Python and (2) TensorFlow.js versions perform in line with the [test results reported in the paper](https://petsymposium.org/popets/2024/popets-2024-0048.php). @atlasharry used the model from Hugging Face that corresponds to our GitHub served model and fixed a small issue in the conversion and also slightly re-tuned the parameters (which, however, did not make much of a difference).

There were a handful of incorrect classifications when testing on the 30+ additional test instances that @dadak-dom created. However, the validity of our original test set stands. The 30+ instances test could have just been an unlucky pick of test set sites or there was something different about those test instances as those were created manually and not according to the prior process.

@PattonYin and @atlasharry, please feel free to include additional information here, if there is anything important to add, to conclude this point. 

## 2. Accuracy Testing

As we know that the model is classifying accurately, we can finally start testing the accuracy of the crawler (by "crawler" I mean the crawler including model, extension and VM).

First, [here is the 100-site test set](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/blob/main/selenium-crawler/test-lists-preparation/100_sites_test_list.csv), and here is [the methodology how the test set was constructed](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/tree/main/selenium-crawler/test-lists-preparation#crawler-performance-test-list-preparation).

### 2.1 Test Set

**Very important**: @atlasharry, @PattonYin, and @ananafrida, if you do any testing on your own, please do not use any of the sites in the test set. The crawler is not allowed to see any test instances to guarantee an unbiased test. If you have used any sites from the test set, we need to randomly replace those sites with unseen ones per the [described methodology](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/tree/main/selenium-crawler/test-lists-preparation#crawler-performance-test-list-preparation).

### 2.2 Two Analyses

As I see it, we need to perform two analyses:

1. Just do a normal evaluation. In other word, just as we ran the extension (with the model) on its own, we just need to run the crawler checking the ground truth against the analysis data.
2. Compare crawler (i.e., crawler including model, extension and VM) results with normal (i.e., just extension/model) results. The point of this exercise is to make a valid claim that our crawler results are actually similar to what a normal user experiences. 

For both steps, we should calculate precision, recall, and F1. 

**Important**: We should only calculate these scores from the positive instances and not the negative instances (i.e., not use weighted scores).

### 2.3 How to Do the Crawler Analysis?

Now, while we have no problem of performing the first analysis (just extension/model) on the same test set run, i.e., evaluating the ground truth and analysis data based on the same test set run, this is naturally not possible for the second analysis (i.e., running the crawler vs running just the extension/model are necessarily two different runs). So, this brings us naturally to the problem how can we distinguish natural fluctuations in site loads in different runs from differences caused by adding our crawler infrastructure?

I think if we run each test, say, three times, we get a sense of the natural differences of each type of run and can distinguish those from the crawler-created differences. In other words, run just the extension/model three times on the test set and check their differences. Then, run the crawler three times and check their differences. Now, averaging out the extension/model runs and the crawler runs, do these averaged runs look very different from one another. So far the theory. I hope it works in practice. 😄 

Now, next question: How can we get a good comparison? We would need to be physically in a place of a VM location and run the extension/model as a normal user. Since I am at a workshop in the Bay Area on November 7, I can do that. I can run just the extension/model (without crawler and VM) and collect the ground truth data (and extension data) as a real California user. Since that is still a few weeks out, the November 7 date should not stop us from already doing the crawl when we are ready. We probably have a sufficient intuitive understanding of whether the accuracy is good. Even if the November 7 test turns out to be not good, we can still re-crawl. 10K sites is "small big data" and can be done fairly quickly.

**Comment by SebastianZimmeck** [comment]
@atlasharry and @PattonYin found that while the model works as expected, there is still an issue with the analysis performed by the extension, i.e., the model results do not seem to be properly processed by the extension.

We discussed that @PattonYin and @atlasharry will:

1. Run the analysis on the complete test set and compare the results against ground truth results to determine the magnitude of the issue. So far there has only been one instance that is not processed correctly. Are there more?
2. The results from task 1 will tell us to which extent we should address the issue, if at all. If there are more instances of incorrect processing by the extension, the set of instances can help us to pinpoint the issue. E.g., what do these instances have in common?

**Comment by SebastianZimmeck** [comment]
@PattonYin and @atlasharry found that the incorrect results produced by the extension need to be addressed indeed. It is a bigger issue.

There are two points to note:
1. The classification suffers from recall; precision is not a problem
2. Errors occur more when numbers are involved (latitude, longitude, ZIP code) as opposed to alphabetical character strings (region, city)

@atlasharry and @PattonYin will look into the extension code, e.g., logging the process at various stages. @PattonYin mentioned that the search functionality may not run correctly.

**Comment by SebastianZimmeck** [comment]
@atlasharry and @PattonYin have resolved the issue of the incorrect classifications.

The issue was caused by an incorrect implementation of escape sequences. In particular, there was one line of code in the pre-processing of the HTTP messages that added multiple backslashes to large parts of a message. This pre-processing happened right before a message was passed to the model for classification and caused mis-classifications.

It is not fully clear why this issue did not come up during the Privacy Pioneer extension testing for the PETS paper. One reason could be that the test instances did not have (as much) escaping as our current test set.

@PattonYin, can you link the file and line of code here?

@atlasharry and @PattonYin, please also add relevant details here, if any.

**Comment by SebastianZimmeck** [comment]
We will proceed as follows:

1. Implement the pre-processing fix in [both extension and crawler](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler?tab=readme-ov-file#7-porting-changes-from-privacy-pioneer-to-the-privacy-pioneer-web-crawler) 
2. Create a new test set. @atlasharry will also add to the [test set creation protocol](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/tree/56f9901bc04a36d613407d90650bd28afa305147/selenium-crawler/test-lists-preparation) how we selected the sites for the new test set
3.  Once we have the test set, we start with the [first test phase](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-2365335466) (under "Just do a normal evaluation. ...). For that, we need to crawl the sites and also check manually the underlying data of the test run to tell whether the classifications are correct. We should also save the data (i.e., HTTP messages) for later reference.

This week, we will do 1 and 2 and prepare 3. So that by Friday next week we can do 3.

**Comment by PattonYin** [comment]
This 

> @atlasharry and @PattonYin have resolved the issue of the incorrect classifications.
> 
> The issue was caused by an incorrect implementation of escape sequences. In particular, there was one line of code in the pre-processing of the HTTP messages that added multiple backslashes to large parts of a message. This pre-processing happened right before a message was passed to the model for classification and caused mis-classifications.
> 
> It is not fully clear why this issue did not come up during the Privacy Pioneer extension testing for the PETS paper. One reason could be that the test instances did not have (as much) escaping as our current test set.
> 
> @PattonYin, can you link the file and line of code here?
> 
> @atlasharry and @PattonYin, please also add relevant details here, if any.

Sure, we added a preprocessing step to the input with the following line of code: 
`const input_cleaned = input.replace(/\\+\"/g, '\\"');`

The key issue we identified is that the number of backslashes preceding the double quotation mark (`"`) directly affects the model's performance. As illustrated in the example, the second snippet containing **2 backslashes + quotation mark** `\\"` is tokenized differently (adding token 1032) compared to the case with **1 backslash + quotation mark** `\"`, leading to different predictions. To address this, the most straightforward solution is to replace *any number of backslashes + quotation mark** with **1 backslash + quotation mark**. 

The explanation for this behavior lies in how the backslash is used as an escape character. When a quotation mark needs to appear within a string, the backslash signals that the quotation mark is part of the string rather than an indicator of its end. Additionally, when this string is saved to files like JSON, another backslash is added as part of the escape sequence. Therefore, **two backslashes** are used to represent a single backslash in the stored string, compounding the number of backslashes introduced.

Since these additional backslashes don't alter the meaning of the text, they can be safely removed during preprocessing.


<img width="383" alt="aea10284e1aaaa3ad9a6dc9f7a4e343" src="https://github.com/user-attachments/assets/c9ecb7a9-9418-4d49-8037-847c8c7ad152">
<img width="452" alt="4d09d193e4d0d7be2f870af9ae7bde6" src="https://github.com/user-attachments/assets/4cf75283-8af6-4d4e-88f4-18837d4e6d44">

**Comment by atlasharry** [comment]
[Here](https://github.com/privacy-tech-lab/privacy-pioneer/commit/35760617a5f1e7f3cacbf9413e085b013329d0d6) is the updated Version with the input clean-up in the Privacy Pioneer. 

Adding to Patton's point, different numbers of backslashes would result in different outputs in the tokenizer. For example:
For the string literals in the snippet: `\"` would result a tokenized input `[1032 1000]`, while `\\\" `would be `[1032 1032 1032 1000]` This difference resulted by incorrect implementation of escape sequences could possibly shift the model's attention or change the importance that assigns to certain parts of the input.

**Comment by atlasharry** [comment]
[This](https://docs.google.com/spreadsheets/d/1b0-cIQumUSuEMLWGLvxxaWYA1DTkrG2rVXnU4SXeQzI/edit?gid=0#gid=0) is the sheet that we use to track the second crawl analysis, we haven't calculated the scores yet, but some intermediary results/observations we have is that:

Among those sites that collect the user's location, the crawler+extension performs pretty well on prediction. In the few cases that lead to a false negative is that 

- The location information extensions chosen to match(call from ipinfo API) do not usually match the sites' location information. E.g. Ipinfo identifies the VM's zip code as 90009, while some sites identify the zip code as 90060. 
-  For sites that use more than one API to collect the users' location, the extension can only identify one among all. (we are still investigating why this happened)

We will continue doing the score calculations and analysis.

**Comment by SebastianZimmeck** [comment]
Thanks, @atlasharry (and @PattonYin)!

A few suggestions:
- Can you identify in the Sheet what the ground truth values are and what the analysis results are?
- Can you include in the Sheet logic for calculating precision, recall, and F1?
- Generally, it is not very clear what is what in the Sheet. Maybe, add a tab with explanations and/or more meaningful tab/column/row/etc. names.

Ideally, we want to have the complete results by Tuesday. 

Also, if you can prepare the protocol for next Wednesday when I am in California for the second part of the test, that would be good. At that time, we have to be sure it works, which is why it would be good it you also test it yourself (from CT; can somebody not knowing about the details of your testing follow it and run the test?).

**Comment by SebastianZimmeck** [comment]
As there were still some incorrect results @PattonYin and @atlasharry identified the following issue as @PattonYin describes:

> Sebastian Zimmeck, we checked the data flow in the extension, and found that the LengthHeuristic and the reqURL is the main cause. 
>  
> Length: 
> According to the existing code and paper, we won't analyze any http message exceeding 100,000 characters. And our analysis revealed that, in many FN cases, the location is not identified because the message containing it exceeds the 100,000-character limit.
> To fix this, we're thinking about changing the heuristic a bit by having it analyzing the first 100,000 characters of the message. 
> But I'm a little worried on the "consistency" issue, since according to the paper, we will directly skip such message.
> 
> reqURL:
> However, removing that heuristic doesn't resolve the issue. Although the snippets are now extracted, the model predicts false frequently. 
> We found this is because of the reqURL appended at the beginning of the text snippet. After removing the reqUrl, the model predicts correctly. (Please Check Image 1)
> 
> Because of this, we're thinking about 2 potential fixes, 1. modify the lengthHeuristic so that it will still analyze the first 100,000 characters and 2. remove the reqUrl in the beginning. Only when both are implemented, the model successfully identifies the "region". (Please check image 2)
 > 
> Just want to confirm if these fixes are acceptable. If these 2 fixes are okay, we can run a crawler test immediately, and see if that improves the model performance.
 > 
> Additional Note: I belive we can safely remove the reqUrl because: this string came from "JSON.stringnify" (please check image 3), which have nothing to do with the snippet itself. 
>  
> 
![image](https://github.com/user-attachments/assets/083e85a6-3329-4299-b587-a35d3ab9eed7)
> 
> ![image](https://github.com/user-attachments/assets/f7803876-ddc0-48c7-baf0-af561a5b5090)
> 
>  ![image](https://github.com/user-attachments/assets/2e6d0012-ec63-4d30-9163-ad219cb9adc3)

Thus, we decided to fix these two issues.

**Comment by SebastianZimmeck** [comment]
Per @atlasharry, we make the following modification:

> Hi Sebastian Zimmeck, we have done some initial crawl with these new changes and we realized removing the requestUrl would indeed causes some unintended consequences. For example, for a reuqestUrl "https://recsys-engine-prd.smiles.com.br/v2/recommendation?place=home_1&id=59fdc7db-8ab9-40bf-8645-74d9c7ed4591&lat=41.5551982&lon=-72.6519254&trace_id=82f83e85-5e39-42a6-92e8-c3f6ba9a0154&request_id=b65f013e-2aa0-411f-a92a-e66156dfb4fa" the Url itself contains the lon and lat. Removing the Url would skip some location information. 
> To fix that, we uses a seemly redundant but useful way to seperately feed reqestUrl, requestBody, responsedata, and a combination of all three(the combined data of all three is how the extension originally did) to the model. In this way, we have fix the original issue and avoid unintended consequences by providing redundancy. After this change, we also tested on the websites that originally extension failed to indetifies location informations, they are now all good.
> 
> ![image](https://github.com/user-attachments/assets/8bcb9a0b-a700-4e34-a3c4-601ffe4b4c38)


Per @atlasharry, we also make the following improvement:

> Another potential improvement we could have is to update the model by the new model I trained in the beginning of the semester. The original model seems to be a little overfitting the training set due to the training parameter and steps they set. Therefore, in cases like this picture shows, if the snippet has some irrelevent data that contains mostly ramdom characters( which it never sees in the training set), our original model would classifies as False. The new model would not have this problem. We have tested more than 10 FN snippets which the original model failed to identifies, the new model all correctly classifies them.
> 
> ![图像](https://github.com/user-attachments/assets/483d5c21-4ded-4523-a22a-b9000c82316c)

**Comment by SebastianZimmeck** [comment]
By next week @PattonYin and @atlasharry will perform the analysis as described under [2.2.1 above](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-2365335466) and present the performance results.

@atlasharry made the most excellent observation that we may not need to perform the analysis under 2.2.2 because PP's mechanism for identfying all location types (lat, ZIP, city, etc) is reliant on the IP address and that mechanism works the same way in every geographic location regardless of the specific IP address. Whether or not PP uses a VM IP address or real IP address does not matter.

**Comment by SebastianZimmeck** [comment]
We are getting close to the end of the testing! Here is where we stand:

- **We do not need a new test set**. The question came up because we made some [modifications](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-2459807773). However, those are modifications unrelated to the models. So we are not tuning our models to the data and, thus, do not need a new test set.
- **We do not need a geographical test** ([above](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-2365335466) "Compare crawler (i.e., crawler including model ...")). The reason is [@atlasharry's point](https://github.com/privacy-tech-lab/privacy-pioneer-web-crawler/issues/9#issuecomment-2479152154). 
- **The classification performance for lat/long, city, and region is good**. For ZIP code we have some discrepancies due to those not matching the ZIP code of the user. We accept this as an inherent limitation.

@PattonYin and @atlasharry will do the following by next week:

- Analyze lat/long performance in one unified performance analysis as opposed to individual lat and long analyses. 
- Write a comment here with a table of the complete performance results for locations. This comment also includes links, e.g., to the Google Colab or other documents where the results are calculated, links to the testing protocol, ... , in short, everything related to this testing.
- Perform tests on non-location items, that is, [monetization and tracking items](https://github.com/privacy-tech-lab/privacy-pioneer?tab=readme-ov-file#7-privacy-practice-analysis).

**Comment by atlasharry** [comment]
We have done some small-scale crawls with a random 344 sites from the tranco-1m list. 

- Total time spent around 8.1  hours, with an average speed of 43 sites per hour
- 119 sites failed: with 76 sites reaching the error page; 19 sites InsecureCertificateError; 6 sites human check error, 7 sites TimeoutError; 10 sites WebDriverError;  and one site NoSuchFrameError. 

After a manual inspection of 119 failed websites, we verified that 95 of them are indeed cannot be opened our human check(The websites' fault, but not crawlers).  
After we excluded those websites, we calculated that our crawler has a success rate of 90.3%

| Category                                   | Details         |
|-------------------------------------------|-----------------|
| Total sites crawled                       | 344 sites       |
| Total time spent                          | 8.1 hours       |
| Average speed                             | 43 sites/hour   |
| Total failures                            | 119 sites       |
| Failures by type: Error page (not our problem) | 76 sites        |
| Failures by type: InsecureCertificateError | 19 sites        |
| Failures by type: Human check error       | 6 sites         |
| Failures by type: TimeoutError            | 7 sites         |
| Failures by type: WebDriverError          | 10 sites        |
| Failures by type: NoSuchFrameError        | 1 site          |

**Comment by PattonYin** [comment]
- We manually visited each website using our own browser and found many sites are **unavailable even without crawler**  (i.e. accessed directly with our browser), so we excluded those in the computation of success rate.
- We tested on both Los Angeles VM and Sydney VM, and here's the data of our investigation

### For LA
| Categories | # of errors | # of Sites Unavailable Without Crawler |
|WebDriverError: Reached Error Page | 76 | 74 |
|InsecureCertificateError | 19 | 17 |
|HumanCheckError | 6 | 3 |
|TimeoutError | 7 | 2 |
|WebDriverError | 10 | 0 |
|NoSuchFrameError | 1 | 0 |

Number of true error instances: 24
Excluding the website with error accessing, the number of total remaining site 344 - 95 = 249
Success rate: (249 - 24) / 249  = 90.36%

### For Sydney
| Categories | # of errors | # of Sites Unavailable Without Crawler |
|WebDriverError: Reached Error Page | 36 | 36 |
|InsecureCertificateError | 7 | 5 |
|HumanCheckError | 4 | 4 |
|TimeoutError | 4 | 1 |

Number of true error instances: 5
Excluding the website with error accessing, the number of total remaining site 141 - 51= 90
Success rate: (90 - 5) / 90 = 94.44%


For more information like the specific URLs for sites that are unavailable / available without crawler, please check the driver: https://drive.google.com/drive/u/1/folders/1U704n1Qne9RvSjr_be81L4oTkAtwj-s-

**Comment by SebastianZimmeck** [comment]
Good results, @atlasharry and @PattonYin!

Are we logging the different errors, in particular, `WebDriverError: Reached Error Page`? I do not think we necessarily need to. I am just asking. We can use your results to extrapolate (or do a similar analysis once we have all the crawl data).

The [testing procedure](https://docs.google.com/document/d/1OBT3AlhU-tCBpd1Jzh1TXoIX-cdlyk_i1K0AWWdYMew/edit?usp=sharing) is good as well! 

I will do some spot checks for the analysis results.

**Comment by SebastianZimmeck** [comment]
As discussed, @atlasharry and @PattonYin, it would be good if you can add in the Labels Sheet the # Count of Uniques Sites with at Least One Positive Ground Truth Instances.

Also, I spot-checked the following results, could you clarify?

1. For amplitude
   
   <img width="432" alt="Screenshot 2024-12-27 at 21 15 38" src="https://github.com/user-attachments/assets/860af81c-cce2-467d-a0a3-044b78f7fd12" />
   
   I see one city (Los Angeles) TP and one city FN in the Labels sheet. The Privacy Pioneer Analysis results ([entries 1-50.csv](https://drive.google.com/file/d/1RoO4A4Oe3hvkqupa8NFK87on5CEjni_3/view?usp=drive_link)) contain only one city (row id 284) while the ground truth results ([amplitude har file](https://drive.google.com/file/d/1Sof_8C7cCxyRm2xWsQo0ObBjucGzrPKO/view?usp=drive_link)) contain three city entries (rows 50250, 66403, 67696). How do those results lead to one city TP and one city FN?

2. For ninelineapparel

   <img width="396" alt="Screenshot 2024-12-27 at 21 28 17" src="https://github.com/user-attachments/assets/24d33a81-6c3e-46e1-b565-e4a3dec591a9" />

   I see one lat/long (34..., -118...) TP and one lat/long FN in the Labels sheet. The Privacy Pioneer Analysis results ([entries last 50.csv](https://drive.google.com/file/d/1RoO4A4Oe3hvkqupa8NFK87on5CEjni_3/view?usp=drive_link)) contain six lat/long rows (row ids 572, 573, 575, 576, 577, 578) while the ground truth results ([ninelineapparel har file](https://drive.google.com/file/d/1ZZol-sqRjdygmDBf7i2_dwk42N0r_DSC/view?usp=drive_link)) contains two lat/long entries (rows 58209 and 68718). How do those results lead to one lat/long TP and one lat/long FN?


3. For sites.toro

   <img width="309" alt="Screenshot 2024-12-27 at 21 42 51" src="https://github.com/user-attachments/assets/865f4c50-f595-408a-8254-9f7a768f9f1c" />

   I see one region (California) TP in the Labels sheet. The Privacy Pioneer Analysis results ([entries last 50.csv](https://drive.google.com/file/d/1RoO4A4Oe3hvkqupa8NFK87on5CEjni_3/view?usp=drive_link)) contain two region rows (ids 1216 and 1217) while the ground truth results ([sites.toro har file](https://drive.google.com/file/d/1YvylmuYUnwi6eC450OHrUcYibLjiGf93/view?usp=drive_link)) contain one California entry (row 9726); a third party request with the region California to termly. How do those results lead to one region TP?

For all three sites, I am trying to understand how you arrived at your Labels sheet results based on the counts in the ground truth and analysis data (e.g., are these instance counts vs site counts, how is first party vs third party playing into this, ...). If you could clarify how you arrived at the Labels sheet results in these three cases, that would be good.

**Comment by atlasharry** [comment]
> As discussed, @atlasharry and @PattonYin, it would be good if you can add in the Labels Sheet the # Count of Uniques Sites with at Least One Positive Ground Truth Instances.
> 
> Also, I spot-checked the following results, could you clarify?
> 
> 1. For amplitude
>    <img alt="Screenshot 2024-12-27 at 21 15 38" width="432" src="https://private-user-images.githubusercontent.com/11034003/399004229-860af81c-cce2-467d-a0a3-044b78f7fd12.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUzOTM3NTgsIm5iZiI6MTczNTM5MzQ1OCwicGF0aCI6Ii8xMTAzNDAwMy8zOTkwMDQyMjktODYwYWY4MWMtY2NlMi00NjdkLWEwYTMtMDQ0Yjc4ZjdmZDEyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI4VDEzNDQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTAzNWJiZjQ0ZTRiYTZmNzMxNzlmZTE2NDljYzZmOTUxOWRkMzNmYTY0NTM3MDBlYzA5OTljZTgyZTU3ZDVmZTEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.2NZwUzzwsxkjX3tJ3KTOT4401HxMGFlZJLGTYQ3MQuo">
>    I see one city (Los Angeles) TP and one city FN in the Labels sheet. The Privacy Pioneer Analysis results ([entries 1-50.csv](https://drive.google.com/file/d/1RoO4A4Oe3hvkqupa8NFK87on5CEjni_3/view?usp=drive_link)) contain only one city (row id 284) while the ground truth results ([amplitude har file](https://drive.google.com/file/d/1Sof_8C7cCxyRm2xWsQo0ObBjucGzrPKO/view?usp=drive_link)) contain three city entries (rows 50250, 66403, 67696). How do those results lead to one city TP and one city FN?
> 2. For ninelineapparel
>    <img alt="Screenshot 2024-12-27 at 21 28 17" width="396" src="https://private-user-images.githubusercontent.com/11034003/399005862-24d33a81-6c3e-46e1-b565-e4a3dec591a9.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUzOTM3NTgsIm5iZiI6MTczNTM5MzQ1OCwicGF0aCI6Ii8xMTAzNDAwMy8zOTkwMDU4NjItMjRkMzNhODEtNmMzZS00NmUxLWI1NjUtZTRhM2RlYzU5MWE5LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI4VDEzNDQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTU2NDc1YjAzZjNkMGM5ZGM4NmJmMTU3NGNiMWUyYjY2YTJiNTAxZjE3NDk2NGRkNjdkMTFiMzI4Y2UxNmIyNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.J6LWEE8YuJi7onlwi32bzfVImhzripbcfpPKByJh8gc">
>    I see one lat/long (34..., -118...) TP and one lat/long FN in the Labels sheet. The Privacy Pioneer Analysis results ([entries last 50.csv](https://drive.google.com/file/d/1RoO4A4Oe3hvkqupa8NFK87on5CEjni_3/view?usp=drive_link)) contain six lat/long rows (row ids 572, 573, 575, 576, 577, 578) while the ground truth results ([ninelineapparel har file](https://drive.google.com/file/d/1ZZol-sqRjdygmDBf7i2_dwk42N0r_DSC/view?usp=drive_link)) contains two lat/long entries (rows 58209 and 68718). How do those results lead to one lat/long TP and one lat/long FN?
> 3. For sites.toro
>    <img alt="Screenshot 2024-12-27 at 21 42 51" width="309" src="https://private-user-images.githubusercontent.com/11034003/399007807-865f4c50-f595-408a-8254-9f7a768f9f1c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzUzOTM3NTgsIm5iZiI6MTczNTM5MzQ1OCwicGF0aCI6Ii8xMTAzNDAwMy8zOTkwMDc4MDctODY1ZjRjNTAtZjU5NS00MDhhLTgyNTQtOWY3YTc2OGY5ZjFjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjI4VDEzNDQxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWVhZThjYTkwNTFjZWI1MGMyYTkyMzVlZjg4NzQxYTNlZjBjYjJmOTA0YWMzMjgzNjA0NTMzMDVlYzk0MTcxZDcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.S5VD-hTSSleycS56cfBaCi_--zikO0VhOjTNdkMPDGY">
>    I see one region (California) TP in the Labels sheet. The Privacy Pioneer Analysis results ([entries last 50.csv](https://drive.google.com/file/d/1RoO4A4Oe3hvkqupa8NFK87on5CEjni_3/view?usp=drive_link)) contain two region rows (ids 1216 and 1217) while the ground truth results ([sites.toro har file](https://drive.google.com/file/d/1YvylmuYUnwi6eC450OHrUcYibLjiGf93/view?usp=drive_link)) contain one California entry (row 9726); a third party request with the region California to termly. How do those results lead to one region TP?
> 
> For all three sites, I am trying to understand how you arrived at your Labels sheet results based on the counts in the ground truth and analysis data (e.g., are these instance counts vs site counts, how is first party vs third party playing into this, ...). If you could clarify how you arrived at the Labels sheet results in these three cases, that would be good.

Hi @SebastianZimmeck, I was in charge of analyzing the last sites you mentioned "sites.toro" and Patton was in charge of "ninelineapparel", and "amplitude".

- For the "site.toro", I think PP only identifies one region row 1216 while 1217 row belongs to another website "bristol". And therefore I identified it as 1 TP. Can you take a look if that is the case on your end?
![image](https://github.com/user-attachments/assets/bde3d9df-dfc8-4124-a393-f78c70ff64ea)

**Comment by PattonYin** [comment]
Hi @SebastianZimmeck. 

- For the "amplitude.com", I'm sorry that this is a mistake I made. When I was labeling the groundtruth for this website, I mistakenly took the last appearacne of the city (api2.amplitude.com/2/httpapi) as an entry in the header. After checking, it is indeed coming from the postData. I'll add 1 unit to the FN column.

- For the "ninelineapparel.com". For the rows with id 572, 573, 575, 576, they are referring to "region" and "city" location collection, only row 577 and row 578 are "fineLocation" and "courseLocation". And since row 577 and 578 are referring to the same snippet from the same third party, that is counted as 1 Positive case, which corresponds to the "1" in TP column.   
![image](https://github.com/user-attachments/assets/30578285-6400-4da5-992c-006d676f244a)

**Comment by SebastianZimmeck** [comment]
> For the "ninelineapparel.com". For the rows with id 572, 573, 575, 576, they are referring to "region" and "city" location collection, only row 577 and row 578 are "fineLocation" and "courseLocation". And since row 577 and 578 are referring to the same snippet from the same third party, that is counted as 1 Positive case, which corresponds to the "1" in TP column.

OK, thanks, @PattonYin. To add for completion, the FN entry in the results is correct because in line 58,209 of the ninelineapparels site, there is another entry (in JavaScript; within the 100,000 characters that is analyzed by Privacy Pioneer) that was not flagged by Privacy Pioneer, possibly, because it is just -118.2, i.e., not very precise.

> For the "site.toro", I think PP only identifies one region row 1216 while 1217 row belongs to another website "bristol". And therefore I identified it as 1 TP. Can you take a look if that is the case on your end?

Yes, you are absolutely right, @atlasharry.

**Comment by SebastianZimmeck** [comment]
As to my question on the [contents of the test directory](https://drive.google.com/drive/folders/1gmpiit6AKXFYp5ryz6GMyASlKqX4Rvru?usp=sharing):

> Also, what do the following files contain:
>  
> - allev 1-50.csv
> - allev_last50.csv
> - analysis_last50.csv
 
@atlasharry clarified that:

> allev is the csv containing all of the evidences, and analysis is the csv file including both data in allev.csv and entries.csv (analysis is the database name).

**Comment by SebastianZimmeck** [comment]
We can work on the spot-checking in the summer during the paper preparation.

**Comment by SebastianZimmeck** [comment]
For the paper, we are only using rule-based monetization classification, i.e., Advertising, Analytics, and Social (per the Disconnect list), for which we have reliable performance results.
